lib/srv/discovery/fetchers/aws-sync/kms.go

```diff
@@ -0,0 +1,214 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package aws_sync
+
+import (
+	"context"
+	"sync"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/kms/types"
+	"github.com/gravitational/trace"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	pb "github.com/gravitational/teleport/gen/proto/go/accessgraph/v1alpha"
+)
+
+type kmsClient interface {
+	kms.ListKeysAPIClient
+	kms.ListResourceTagsAPIClient
+	kms.ListAliasesAPIClient
+	DescribeKey(ctx context.Context, params *kms.DescribeKeyInput, optFns ...func(*kms.Options)) (*kms.DescribeKeyOutput, error)
+	GetKeyPolicy(ctx context.Context, params *kms.GetKeyPolicyInput, optFns ...func(*kms.Options)) (*kms.GetKeyPolicyOutput, error)
+}
+
+// pollAWSKMSKeys is a function that returns a function that fetches
+// AWS KMS keys and their inline and attached policies.
+func (a *Fetcher) pollAWSKMSKeys(ctx context.Context, result *Resources, collectErr func(error)) func() error {
+	return func() error {
+		var err error
+		result.KMSKeys, err = a.fetchKMSKeys(ctx)
+		if err != nil {
+			collectErr(trace.Wrap(err, "failed to fetch kms"))
+		}
+		return nil
+	}
+}
+
+func (a *Fetcher) fetchKMSKeys(ctx context.Context) ([]*pb.AWSKMSKeyV1, error) {
+	var keys []*pb.AWSKMSKeyV1
+	var errs []error
+	var mu sync.Mutex
+
+	collectKeys := func(nextKeys []*pb.AWSKMSKeyV1, nextErrs ...error) {
+		mu.Lock()
+		defer mu.Unlock()
+		if len(nextErrs) > 0 {
+			errs = append(errs, nextErrs...)
+		}
+		if nextKeys != nil {
+			keys = append(keys, nextKeys...)
+		}
+	}
+
+	errGroup := errgroup.Group{}
+	// Set the limit to 5 to avoid too many concurrent requests.
+	// This is a temporary solution until we have a better way to limit the
+	// number of concurrent requests.
+	errGroup.SetLimit(5)
+
+	for _, region := range a.Regions {
+		errGroup.Go(func() error {
+			awsCfg, err := a.AWSConfigProvider.GetConfig(ctx, region, a.getAWSOptions()...)
+			if err != nil {
+				collectKeys(nil, trace.Wrap(err))
+				return nil
+			}
+			client := a.awsClients.getKMSClient(awsCfg)
+			a.collectKMSKeys(ctx, client, collectKeys, region)
```

Comment:
Can we just return the keys from `a.collectKMSKeys` and then call `collectKeys` on it here at the caller side? Passing the collector function into this function feels more confusing than it should be tbh.

Reply:
We can and I have updated tis. 

I've tried to follow the patterns established for other AWS resource fetching. if I return `[]*pb.AWSKMSKeyV1, error` here I need to aggregate potentially multiple errors here, rather than just on the top level but maybe that doesn't matter.